File: board.py
import json
from line import Line, Vector, cleanupColinearTrackPair, splitIntersectingLines, getIslands, makePolyLines, dist, kdboxtree, kdboxinside
from copy import copy
from collections import defaultdict, namedtuple
import math

def floatToString(inputValue):
    return ('%f' % inputValue).rstrip('0').rstrip('.')
def pointToStr(pt):
    return f"{floatToString(round(pt.x + 4000, 5))} {floatToString(round(pt.y + 3000, 5))}"

Transform = namedtuple('Transform', 'pos angle')
Via = namedtuple('Via', 'pos diameter drill')

def Pos(x, y, tx):
    return Vector(float(x) - 4000, float(y) - 3000)

class Board:
    def load(self, filename):
        self.nextTrackId = 0
        self.board = json.load(open(filename, encoding='utf-8'))
        self.loadTracks()
        print("Cleaning up overlapping tracks")
        self.cleanupColinearTracks()
        print("Splitting intersecting tracks")
        self.splitIntersectingLines()
        print("Identifying connected track sections")
        self.assignIslands()

    def save(self, filename):
        self.saveTracks()
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(self.board, f, indent=2, sort_keys=False, ensure_ascii=False)

    def cleanupColinearTracks(self):
        for tracks in self.tracksByNetAndLayer.values():
            #trackLookup = kdboxtree(list((t.bounds(), t) for t in tracks))
            for t in tracks:
                for t2 in tracks: #kdboxinside(trackLookup, t.bounds()):
                    if 1: #id(t) < id(t2): doens't work???
                        if cleanupColinearTrackPair(t, t2, tracks):
                            break

    def splitIntersectingLines(self):
        for tracks in self.tracksByNetAndLayer.values():
            splitIntersectingLines(tracks)

    def assignIslands(self):
        for tracks in self.tracksByNetAndLayer.values():
            getIslands(tracks)

    def getShapeId(self):
        self.highestShapeId += 1
        return 'gge%s' % self.highestShapeId

    def addShape(self, shape):
        self.board['shape'].append(shape)

    def loadTracks(self):
        self.highestShapeId = 0
        self.tracksByNetAndLayer = defaultdict(list)
        self.viasByNet = defaultdict(list)
        self.loadShapes(self.board['shape'])

    def loadShapes(self, shapes, tx=Transform(Vector(0, 0), 0)):
        for shapeIndex, shape in enumerate(shapes):
            shape = shape.split('~')
            if shape[0] == 'LIB':
                x, y, attributes, rotation, importFlag, shapeId, _, _, _, locked = shape[1:11]
                shapeList = '~'.join(shape).split('#@$')[1:]
                rotation = float(rotation or 0)
                self.loadShapes(shapeList, Transform(Pos(x, y, tx), rotation + tx.angle))
            elif shape[0] == 'TRACK':
                width, layer, net, coords, shapeId, locked = shape[1:]
                if layer not in ('1', '2'):
                    continue
                tracks = self.tracksByNetAndLayer[net+'~'+layer]
                coords = [float(c) for c in coords.split(' ')]
                width = float(width)
                start = Pos(coords[0], coords[1], tx)
                for i in range(2, len(coords), 2):
                    end = Pos(coords[i], coords[i + 1], tx)
                    t = Line(start, end)
                    t.shapeIndex = shapeIndex
                    t.id = shapeId
                    t.width = width
                    t.layer = layer
                    t.net = net
                    t.locked = locked
                    start = copy(end)
                    if t.update():
                        tracks.append(t)
            elif shape[0] == 'VIA':
                x, y, diameter, net, drill, shapeId, locked = shape[1:]
                pos = Pos(x, y, tx)
                self.viasByNet[net].append(Via(pos, float(diameter), float(drill)))
            elif shape[0] == 'PAD':
                (shapeType, x, y, w, h, layer, net, num, holeRadius, points, 
                 rotation, shapeId, holeLength, holePoints, plated, locked, 
                 pasteExpansion, solderMaskExpansion, _) = shape[1:]
                pos = Pos(x, y, tx)
                if float(holeRadius):
                    self.viasByNet[net].append(Via(pos, min(float(w), float(h)), float(holeRadius)))
            else:
                for field in shape:
                    if field.startswith('gge'):
                        shapeId = field
                        break
            if shapeId.startswith('gge') and shape[0] != 'LIB':
                numericId = int(shapeId[3:])
                if numericId > self.highestShapeId:
                    self.highestShapeId = numericId

    def saveTracks(self):
        tracksByShape = {}
        for tracks in self.tracksByNetAndLayer.values():
            for track in tracks:
                tracksByShape.setdefault(track.shapeIndex, []).append(track)
        for shapeIndex, tracks in tracksByShape.items():
            track = tracks[0]
            trackId = track.id
            for coords in makePolyLines(tracks):
                if shapeIndex is None:
                    trackId = self.getShapeId()
                shape = '~'.join((
                    'TRACK',
                    floatToString(track.width),
                    track.layer,
                    track.net,
                    ' '.join((pointToStr(c) for c in coords)),
                    trackId,
                    track.locked))
                if shapeIndex != None:
                    self.board['shape'][shapeIndex] = shape
                    shapeIndex = None
                else:
                    self.addShape(shape)

File: node.js/board.js
const { writeFile } = require("fs");
const { Point } = require("paper");
const { Line, assignIslands, makePolyLines } = require("./line.js");

function floatToString(inputValue) {
    return inputValue;
    //return ('%f' % inputValue).rstrip('0').rstrip('.')
}
function pointToStr(pt) {
    return floatToString(pt.x) + " " + floatToString(pt.y);
    //return "${floatToString(round(pt.x + 4000, 5))} ${floatToString(round(pt.y + 3000, 5))}";
}

class Via {
    constructor(pos, diameter, holeRadius) {
        this.pos = pos;
        this.diameter = diameter;
        this.holeRadius = holeRadius;
    }
}
class Board {
    constructor(json) {
        this.nextTrackId = 0;
        this.board = JSON.parse(json);
        this.loadTracks();
        //console.log("Cleaning up overlapping tracks");
        //this.cleanupColinearTracks();
        //console.log("Splitting intersecting tracks");
        //this.splitIntersectingLines();
        console.log("Identifying connected track sections");
        this.assignIslands();
    }

    save(filename) {
        this.saveTracks();
        writeFile(filename, JSON.stringify(this.board, null, 2), ()=>{});
    }

    cleanupColinearTracks() {
        for(const tracks of Object.values(this.tracksByNetAndLayer)) {
            //var trackLookup = kdboxtree(list((t.bounds(), t) for t in tracks))
            for(const t of tracks) {
                for(const t2 of tracks) { // kdboxinside(trackLookup, t.bounds()):
                    if (1) //id(t) < id(t2): doens't work???
                        if (cleanupColinearTrackPair(t, t2, tracks))
                            break;
                }
            }
        }
    }

    splitIntersectingLines() {
        for(const tracks of Object.values(this.tracksByNetAndLayer)) {
            splitIntersectingLines(tracks);
        }
    }
    assignIslands() {
        for(const tracks of Object.values(this.tracksByNetAndLayer)) {
            assignIslands(tracks);
        }
    }

    allocateShapeId() {
        this.highestShapeId += 1;
        return `gge${this.highestShapeId}`;
    }

    addShape(shape) {
        this.board['shape'].push(shape);
    }

    loadTracks() {
        this.highestShapeId = 0;
        this.tracksByNetAndLayer = {};
        this.viasByNet = {};
        this.loadShapes(this.board['shape']);
    }

    loadShapes(shapes) {
        for(const [shapeIndex, shapeStr] of shapes.entries()) {
            const shape = shapeStr.split("~");
            var shapeId;
            if (shape[0] == "LIB") {
                const [, x, y, attributes, rotation, importFlag, _shapeId, , , , locked] = shape;
                shapeId = _shapeId;
                const shapeList = shape.join("~").split("#@$").slice(1);
                this.loadShapes(shapeList);
            } else if (shape[0] == "TRACK") {
                const [, widthStr, layer, net, coordsStr, _shapeId, locked] = shape;
                shapeId = _shapeId;
                if (layer == "1" || layer == "2") {
                    const key = net+'~'+layer;
                    if (!(key in this.tracksByNetAndLayer))
                        this.tracksByNetAndLayer[key] = [];
                    const tracks = this.tracksByNetAndLayer[key];
                    const coords = coordsStr.split(' ').map(Number);
                    const width = Number(widthStr);
                    var start = new Point(coords[0], coords[1]);
                    for(var i = 2; i < coords.length; i += 2) {
                        const end = new Point(coords[i], coords[i + 1]);
                        var t = new Line(start, end);
                        t.shapeIndex = shapeIndex;
                        t.id = shapeId;
                        t.width = width;
                        t.layer = layer;
                        t.net = net;
                        t.locked = locked;
                        start = end;
                        if (t.length > 0)
                            tracks.push(t);
                    }
                }
             } else if (shape[0] == 'VIA') {
                const [, x, y, diameter, net, drill, _shapeId, locked] = shape;
                shapeId = _shapeId;
                const pos = new Point(Number(x), Number(y));
                if (!(net in this.viasByNet))
                    this.viasByNet[net] = [];
                this.viasByNet[net].push(new Via(pos, Number(diameter), Number(drill)));
             } else if (shape[0] == 'PAD') {
                const [, shapeType, x, y, w, h, layer, net, num, holeRadius, points, 
                 rotation, _shapeId, holeLength, holePoints, plated, locked, 
                 pasteExpansion, solderMaskExpansion] = shape;
                shapeId = _shapeId;
                const pos = new Point(Number(x), Number(y));
                if (Number(holeRadius)) {
                    if (!(net in this.viasByNet))
                        this.viasByNet[net] = [];    
                    this.viasByNet[net].push(new Via(pos, Math.min(Number(w), Number(h)), Number(holeRadius)));
                }
            } else {
                for(const field of shape) {
                    if (field.startsWith('gge')) {
                        shapeId = field;
                        break;
                    }
                }
            }
            if (shapeId.startsWith('gge') && shape[0] != 'LIB') {
                const numericId = Number(shapeId.substr(3));
                if (numericId > this.highestShapeId)
                    this.highestShapeId = numericId;
            }
        }
    }

    saveTracks() {
        var tracksByShape = {};
        for(const tracks of Object.values(this.tracksByNetAndLayer)) {
            for(const track of tracks) {
                if (!(track.shapeIndex in tracksByShape))
                    tracksByShape[track.shapeIndex] = [];
                tracksByShape[track.shapeIndex].push(track);
            }
        }
        for(var [shapeIndex, tracks] of Object.entries(tracksByShape)) {
            const track = tracks[0];
            var trackId = track.id;
            for(const coords of makePolyLines(tracks)) {
                if (shapeIndex === null)
                    trackId = this.allocateShapeId();
                const shape = [
                    'TRACK',
                    floatToString(track.width),
                    track.layer,
                    track.net,
                    coords.map(pointToStr).join(' '),
                    trackId,
                    track.locked].join('~');
                if (shapeIndex != null) {
                    this.board['shape'][shapeIndex] = shape;
                    shapeIndex = null;
                } else {
                    this.addShape(shape);
                }
            }
        }
    }
}

module.exports = {Board, pointToStr, floatToString};
